### PR TITLE
Feature/heatmap

### PR DESCRIPTION
**Pull request by Andries-Smit**

### CODE REVIEW

File: src/HeatMap/HeatMap.webmodeler.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        require("../ui/InfoTooltip.scss") +
+        require("plotly.js/src/css/style.scss")
+    );
+}
+
+export function getVisibleProperties(valueMap: HeatMapContainerProps, visibilityMap: VisibilityMap<HeatMapContainerProps>) { // tslint:disable-line max-line-length
+    if (valueMap.dataSourceType === "XPath") {
+        visibilityMap.dataSourceMicroflow = false;
+    } else if (valueMap.dataSourceType === "microflow") {
+        visibilityMap.entityConstraint = false;
```
Comment (Andries-Smit) [inline src/HeatMap/HeatMap.webmodeler.ts:54]: hide `horizontalSortAttribute`,  `horizontalSortOrder`, `verticalSortAttribute`, `verticalSortOrder` for datasource microflow

File: src/HeatMap/components/HeatMap.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        ]);
+    }
+
+    private getTextAnnotations() {
+        const annotations: {}[] = [];
+        if (this.props.data) {
+            for (let i = 0; i < this.props.data.y.length; i++) {
+                for (let j = 0; j < this.props.data.x.length; j++) {
+                    const currentValue = this.props.data.z[ i ][ j ];
+                    const textColor = currentValue !== 0.0 ? "white" : "black";
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMap.ts]: We should not set the color this static. As the user can change color scale.
I would make more sense to have a to calculate the contrast
``` js
function getContrastYIQ(hexcolor){
    var r = parseInt(hexcolor.substr(0,2),16);
    var g = parseInt(hexcolor.substr(2,2),16);
    var b = parseInt(hexcolor.substr(4,2),16);
    var yiq = ((r*299)+(g*587)+(b*114))/1000;
    return (yiq >= 128) ? 'black' : 'white';
}
```

File: src/HeatMap/components/HeatMap.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                    const result = {
+                        xref: "x1",
+                        yref: "y1",
+                        x: this.props.data.x[ j ],
+                        y: this.props.data.y[ i ],
+                        text: this.props.data.z[ i ][ j ],
+                        font: {
+                            family: "Arial",
+                            size: 12,
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMap.ts]: remove font size and family, if possible

File: src/HeatMap/components/HeatMapContainer.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                );
+
+                return zData ? Number(zData.get(this.props.valueAttribute)) : 0;
+            }));
+    }
+
+    private processColorScale(): (string | number)[][] {
+        return this.props.scaleColors.length > 1
+            ? this.props.scaleColors.map(colors => [ Math.abs(colors.valuePercentage / 100), colors.colour ])
+            : [ [ 0, "blue" ], [ 1, "red" ] ];
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: These default are already set in the lib, so we can omit them for now. We might need it we start charts themes

File: src/HeatMap/components/HeatMapContainer.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    private getValues(data: mendix.lib.MxObject[], attribute: string): string[] {
+        const values: string[] = [];
+        if (data.length) {
+            data.forEach(item => {
```
Comment (Andries-Smit) [inline src/HeatMap/components/HeatMapContainer.ts]: no need to create const and check length
just
`return data.map(.....`
Ow now i see , you dont want to add all. 
`return data.reduce` can do that trick

**Review by Andries-Smit — COMMENTED** [review]
Old notes, might still apply

File: src/HeatMap/HeatMap.webmodeler.ts (hunk 1 above)
Comment (edwinmp) [reply src/HeatMap/HeatMap.webmodeler.ts:54]: Done.

File: src/HeatMap/components/HeatMap.ts (hunk 3 above)
Comment (edwinmp) [reply src/HeatMap/components/HeatMap.ts]: Done.

File: src/HeatMap/components/HeatMapContainer.ts (hunk 4 above)
Comment (edwinmp) [reply src/HeatMap/components/HeatMapContainer.ts]: Used the values Danny provided in his feedback.

File: src/HeatMap/components/HeatMap.ts (hunk 2 above)
Comment (edwinmp) [reply src/HeatMap/components/HeatMap.ts]: Couldn't really figure out how to do this, cos I have no hex colour to base on at any point in this code.

File: src/HeatMap/components/HeatMapContainer.ts (hunk 5 above)
Comment (edwinmp) [reply src/HeatMap/components/HeatMapContainer.ts]: My understanding of reduce is that it requires a return for each iteration, which wouldn't really work for me.

**Pull request closed without merging**